### PR TITLE
csclient: allow uploading a specified revision

### PR DESCRIPTION
**Pull request by rogpeppe**
This means we won't need to duplicate most of the UploadCharm
functionality in the charmstoretesting package, and paves the way
for factoring the csclient package out of charmstore.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/921/
Test PASSed.

**Comment by frankban** [comment]
:+1: thank you!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/922/
Test PASSed.

File: csclient/csclient.go [hunk 1]
```diff
@@ -164,8 +164,17 @@ func (c *Client) GetArchive(id *charm.Reference) (r io.ReadCloser, eid *charm.Re
...
+// charm.CharmArchive.
+//
+// The id should include the series. If the revision is omitted,
+// a new revision will be chosen, otherwise the specified revision
+// will be uploaded. In general, specifying a revision should
+// be avoided - it is provided for tests, and will probably be prohibited
+// for non-privileged clients in the future.
+//
+// UploadCharm returns the id that the charm has been given in the
+// store - this will be the same as id except for the revision.
```
Comment (mhilton) [inline csclient/csclient.go]: except if you specified the revision in which case it will be that same as id.

**Comment by mhilton** [comment]
:+1: LGTM

File: csclient/csclient.go (hunk 1 above)
Comment (rogpeppe) [inline csclient/csclient.go]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://localhost:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/923/
Test PASSed.

**Pull request merged**